components/edit-collective/sections/SetupTwoFactorAuth.js

```diff
@@ -0,0 +1,335 @@
+    const options = {
+      name: `${this.props.collectiveName}'s Open Collective`,
+      length: 64,
+    };
+    const secret = speakeasy.generateSecret(options);
+    // speakeasy options object does not add issuer as expected so we add it ourselves to the otp url
+    const issuer = '&issuer=Open%20Collective';
+    const fullOTPUrl = secret.otpauth_url + issuer;
+    this.setState({ secret, base32: secret.base32, otpauth_url: fullOTPUrl });
+  }
+
+  async submit(values) {
+    try {
+      // verify QR code
+      const { TOTPCode } = values;
+      const verified = speakeasy.totp.verify({
+        secret: this.state.base32,
+        encoding: 'base32',
+        window: 1,
+        token: TOTPCode,
+      });
+
+      // if not verified, ask the user to try again
+      if (!verified) {
+        this.setState({ error: 'Secret not verified. Please try again.' });
+      }
+
+      // if ok, send secret to backend
+      if (verified) {
+        this.setState({ submitting: true });
+
+        const account = {
+          id: this.props.data.account.id,
+        };
+
+        await this.props
+          .addTwoFactorAuthToAccount({
+            variables: {
+              account,
+              token: this.state.base32,
+            },
+          })
+          .then(() => {
+            confettiFireworks(2000, { zIndex: 3000 });
+          });
+        this.setState({ tokenAdded: true, error: null, submitting: false });
+      }
+    } catch (err) {
+      const errorMsg = getErrorFromGraphqlException(err).message;
+      this.setState({ error: errorMsg });
+    }
+  }
+
+  render() {
+    const { intl, data } = this.props;
+    const { error, submitting, tokenAdded, secret, base32, otpauth_url } = this.state;
+
+    const { loading } = data;
+
+    if (loading) {
+      return <Loading />;
+    }
+
+    const account = data && data.account;
+    const doesAccountAlreadyHave2FA = tokenAdded || account.hasTwoFactorAuth;
+
+    const initialValues = {
+      TOTPCode: '',
+    };
+
+    const validate = values => {
+      const errors = {};
+
+      if (values.TOTPCode.toString().length > 6) {
+        errors.TOTPCode = intl.formatMessage(messages.errorWrongLength);
+      }
+
+      return errors;
+    };
+
+    return (
+      <Flex flexDirection="column">
+        {error && (
+          <MessageBox type="error" withIcon my={2}>
+            {error}
+          </MessageBox>
+        )}
+        <Flex flexDirection="column" my={2}>
+          <H2>
+            <FormattedMessage id="TwoFactorAuth.Setup.Title" defaultMessage="Set up two-factor authentication" />
+          </H2>
+          {doesAccountAlreadyHave2FA ? (
+            <Flex alignItems="center">
+              <CheckCircle color="green" size="24" />
+              <P ml={2}>
+                <FormattedMessage
+                  id="TwoFactorAuth.Setup.AlreadyAdded"
+                  defaultMessage="Two-factor authentication is enabled on this account. Well done! 🎉"
```

Comment:
We don't use it very often, but the "standard" way to display these kind of messages is supposed to be `<MessageBox type="success" withIcon>`.